cohttp-lwt/src/connection.ml

```diff
@@ -0,0 +1,264 @@
+module Make (Net : S.Net) : (S.Connection with module Net = Net) =
+struct
+  module Net = Net
+  module IO = Net.IO
+  module Response = Make.Response (IO)
+  module Request = Make.Request (IO)
+  module Header = Cohttp.Header
+
+  open IO
+
+  exception Retry = S.Retry
+
+  let src = Logs.Src.create "cohttp.lwt.client" ~doc:"Cohttp Lwt client"
+
+  module Log = (val Logs.src_log src : Logs.LOG)
+
+  type state =
+    | Connecting of (IO.ic * IO.oc) Lwt.t
+    | Full of (IO.ic * IO.oc)
+    | Closing of (IO.ic * IO.oc)
+    | Half of IO.ic
+    | Closed
+    | Failed of exn
+  type req_resr =
+    { req :Request.t
+    ; body :Body.t
+    ; res_r :(Response.t * Body.t) Lwt.u
+    ; res :(Response.t * Body.t) Lwt.t }
+  type persistent = [ `True | `False | `Unknown ]
+  type t =
+    { mutable state :state
+    ; mutable persistent :persistent
+    ; in_flight :req_resr Queue.t
+    ; waiting :req_resr Queue.t
+    ; condition :unit Lwt_condition.t (* watching queues *)
+    ; finalise :(t -> unit Lwt.t)
+    }
+
+  let length connection =
+    Queue.length connection.in_flight +
+    Queue.length connection.waiting
+
+  let notify connection =
+    Lwt_condition.wait connection.condition
+
+  let queue_fail connection q e =
+    Queue.iter (fun {res_r; _} -> Lwt.wakeup_later_exn res_r e) q;
+    Queue.clear q;
+    Lwt_condition.broadcast_exn connection.condition e
+
+  let close connection =
+    match connection.state with
+    | Connecting channels ->
+      connection.state <- Closed;
+      Lwt.cancel channels;
+      Lwt.on_success channels (fun (ic, oc) -> Net.close ic oc);
+      Lwt_condition.broadcast connection.condition ()
+    | Closing (ic, oc)
+    | Full (ic, oc) ->
+      Net.close ic oc;
+      connection.state <- Closed;
+      Lwt_condition.broadcast connection.condition ()
+    | Half ic ->
+      Net.close_in ic;
+      connection.state <- Closed;
+      Lwt_condition.broadcast connection.condition ()
+    | Closed | Failed _ -> ()
+
+  let shutdown connection =
+    match connection.state with
+    | Connecting channels ->
+      Lwt.async @@ fun () ->
+      channels >>= fun channels ->
+      (* need to loose possible race with writer *)
+      Lwt.pause () >>= fun () ->
```

Comment:
Hmm. It seems there is no race anymore. What I was thinking about is this:
```
open Connection
let connection = create endpoint in
let response = request connection request in
shutdown connection
(* now both, shutdown and the request wait for the channels to become ready *)
(* request would fail if shutdown won the race *).
```
But now `request` returns immediately without handling control to Lwt having enqueued the request.
The writer won't be bothered by the `Closing` state, since it sees the full queue.

I may need to have another look. But right now I can't find the race.

Reply:
I would consider that an incorrect use though, I would rather use Lwt.finally and call shutdown there, no?

Reply:
No, I actually want that to work. It allows one to send `EOF` before even reading the response, so the remote end knows we are finished as early as possible.